### ISSUE TITLE
Shell (bash) closing unexpectedly

### ISSUE THREAD

**Issue by tokudan** [direnv/direnv#848]
**Describe the bug**
In some cases, direnv causes the shell to exit unexpectedly for the user.

**To Reproduce**
1. A line like this in your .bashrc can help debugging: `export PS1="SHLVL=${SHLVL}  ${PS1}"`, also set +x is useful to observe.
2. Run (manual or through .bashrc or similar): `eval "$(direnv hook bash)"`
3. Enter the following command and press Enter: `echo $(cat)`
4. Press CTRL+C to abort the cat command
5. cat exits as expected, but the shell also closes, possibly exiting your terminal if it was the top level shell.
This only happens after `direnv hook bash`. If the hook is not installed, the shell is fine.

**Expected behavior**
The shell should have given me a prompt instead of exiting.

**Environment**
 - OS: NixOS 21.05.4386.9b2fd29751d (Okapi)
 - Shell: GNU bash, Version 4.4.23(1)-release
 - Direnv version: 2.28.0

**Additional context**
Transcript:
```
[user@host:~]$ set -x
++ _direnv_hook
++ local previous_exit_status=0
++ trap -- '' SIGINT
+++ /nix/store/jq1d46m14di7wrcmd3rsiy9sf5zxkbfm-direnv-2.28.0/bin/direnv export bash
++ eval ''
++ trap - SIGINT
++ return 0
SHLVL=2
[user@host:~]$ bash
+ bash
SHLVL=3
[user@host:~]$ set -x
++ _direnv_hook
++ local previous_exit_status=0
++ trap -- '' SIGINT
+++ /nix/store/jq1d46m14di7wrcmd3rsiy9sf5zxkbfm-direnv-2.28.0/bin/direnv export bash
++ eval ''
++ trap - SIGINT
++ return 0
SHLVL=3
[user@host:~]$ echo $(cat)
++ cat
^C
++ _direnv_hook
++ local previous_exit_status=130
++ trap -- '' SIGINT
+++ /nix/store/jq1d46m14di7wrcmd3rsiy9sf5zxkbfm-direnv-2.28.0/bin/direnv export bash
++ eval ''
++ trap - SIGINT
++ return 130
SHLVL=2
[user@host:~]$ 
```

**Comment by yermulnik** [comment]
I think this had been fixed with Bash 5.1 release. If you're on the older version of Bash here's what worked for me https://github.com/direnv/direnv/issues/627 (specifically [this](https://github.com/direnv/direnv/issues/627#issuecomment-634504650))

**Comment by tokudan** [comment]
@yermulnik Yep, that's it. The workaround also works. Thanks!
